### ISSUE TITLE
Add traits for generic norms and distances.

### ISSUE THREAD

**Issue by Nuramon27** [rust-num/num-traits#121]
In many iterative numerical algorithms, one has to calculate the distance between a solution candidate and a reference value in order to check for convergence. But currently there is no trait which would enable the computation of such a distance generically. 

There is `Signed` which provides an `abs` function, but this trait can only be implemented for types with a notion of signedness. The correct mathematical concept, however, are norms and metrics, with the latter being reasonable for basically _any_ numeric type.

So I propose adding two traits, `Norm` and `Distance`, each with one method of the same name. The `norm`-function of the `Norm` trait returns the norm of a _single_ value, the `distance`-function of the `Distance` trait returns the distance between _two_ values.

In my branch [dist](https://github.com/Nuramon27/num-traits/tree/dist) I have written a draft of these traits and implemented them for the primitive numeric types.

**Comment by cuviper** [comment]
It's a reasonable idea. Can you open it as a PR for discussion?

One thing to consider is that there are multiple possible definitions of norm, especially when more dimensions are added like in `Complex<T>`. e.g. [Euclidean norm](https://en.wikipedia.org/wiki/Norm_(mathematics)#Euclidean_norm) vs. [Manhattan norm](https://en.wikipedia.org/wiki/Norm_(mathematics)#Taxicab_norm_or_Manhattan_norm).

**Comment by JohnScience** [comment]
> It's a reasonable idea. Can you open it as a PR for discussion?
> 
> One thing to consider is that there are multiple possible definitions of norm, especially when more dimensions are added like in `Complex<T>`. e.g. [Euclidean norm](https://en.wikipedia.org/wiki/Norm_(mathematics)#Euclidean_norm) vs. [Manhattan norm](https://en.wikipedia.org/wiki/Norm_(mathematics)#Taxicab_norm_or_Manhattan_norm).

Are you sure that it must be the responsibility of num-traits?
Creating another managed crate (for example, "num-maths-traits") which would contain category-, group-, field-, ring-, space-, and other more general traits sounds a better idea. In this case, trait inheritance from more general traits can be an opt-in feature enabled by default.

The problem is not even that num-traits may end up becoming too of a heavy dependency in attempts to fit all needs but that num-traits' purpose statement ("Numeric traits for generic mathematics in Rust"), would, in this case, become unclear to me and, perhaps, someone else.

**Comment by cuviper** [comment]
> Are you sure that it must be the responsibility of num-traits?

Not at all. If there are interested parties like yourself who want to explore this in a new crate (and not get bogged down in `num-traits`' conservative API), then feel free!
